File: src/checkincludes.rs
use std::path::{Path, PathBuf};
use std::cmp::Ordering;
use std::collections::HashSet;

use structopt::StructOpt;
use regex::Regex;


use crate::
{
    core,
    builddata,
    printer,
    rust
};


#[derive(StructOpt, Debug)]
pub struct MainData
{
    /// Files to look at
    pub files: Vec<PathBuf>,

    /// Print general file status at the end
    #[structopt(long)]
    pub status: bool,

    /// Use verbose output
    #[structopt(long)]
    pub verbose: bool,
}


#[derive(StructOpt, Debug)]
pub enum Options
{
    /// Print headers that don't match any pattern so you can add more regexes
    MissingPatterns
    {
        #[structopt(flatten)]
        main: MainData
    },
    
    /// Print headers that can't be fixed
    ListUnfixable
    {
        #[structopt(flatten)]
        main: MainData,

        /// Print all errors per file, not just the first one
        #[structopt(long)]
        all: bool
    },


    /// Check for style errors and error out
    Check
    {
        #[structopt(flatten)]
        main: MainData
    },

    /// Fix style errors and print unfixable
    Fix
    {
        #[structopt(flatten)]
        main: MainData,

        /// Write fixes to file
        #[structopt(long)]
        write: bool
    }
}



struct Include
{
    line_class: i32,
    line: String
}

impl Include
{
    fn new(c: i32, l: &str) -> Include
    {
        Include
        {
            line_class: c,
            line: l.to_string()
        }
    }
}


fn include_compare(lhs: &Include, rhs: &Include) -> Ordering
{
    if lhs.line_class == rhs.line_class
    {
        lhs.line.cmp(&rhs.line)
    }
    else
    {
        lhs.line_class.cmp(&rhs.line_class)
    }
}

fn error(print: &mut printer::Printer, filename: &Path, line: usize, message: &str)
{
    print.error(format!("{}({}): error CHK3030: {}", filename.display(), line, message).as_str());
}

fn warning(print: &mut printer::Printer, filename: &Path, line: usize, message: &str)
{
    print.warning(format!("{}({}): warning CHK3030: {}", filename.display(), line, message).as_str());
}

#[derive(Debug, Copy, Clone)]
enum MessageType
{
    Error, Warning
}

fn print_message(message_type: MessageType, print: &mut printer::Printer, filename: &Path, line: usize, message: &str)
{
    match message_type
    {
        MessageType::Error   => error(print, filename, line, message),
        MessageType::Warning => warning(print, filename, line, message)
    }
}

pub fn get_replacer(file_stem: &str) -> core::TextReplacer
{
    let mut replacer = core::TextReplacer::new();
    replacer.add("{file_stem}", file_stem);
    replacer
}


fn classify_line
(
    missing_files: &mut HashSet<String>,
    print: &mut printer::Printer,
    data: &builddata::BuildData,
    line: &str,
    filename: &Path,
    line_num: usize
) -> Option<i32>
{
    let replacer = get_replacer(filename.file_stem().unwrap().to_str().unwrap());

    for (index, included_regex_group) in data.includes.iter().enumerate()
    {
        for included_regex in included_regex_group
        {
            let re = match included_regex
            {
                builddata::OptionalRegex::Dynamic(regex) =>
                {
                    let regex_source = replacer.replace(regex);
                    match Regex::new(&regex_source)
                    {
                        Ok(re) => { re },
                        Err(err) =>
                        {
                            print.error(format!("{} -> {} is invalid regex: {}", regex, regex_source, err).as_str());
                            return None;
                        }
                    }
                },
                builddata::OptionalRegex::Static(re) => {re.clone()},
                builddata::OptionalRegex::Failed(err) =>
                {
                    print.error(err.as_str());
                    return None;
                }
            };

            if re.is_match(line)
            {
                return Some(index.try_into().unwrap());
            }
        }
    }

    if missing_files.contains(line) == false
    {
        missing_files.insert(line.to_string());
        error(print, filename, line_num, format!("{} is a invalid header", line).as_str());
    }
    None
}


fn get_text_after_include_quote(line: &str) -> String
{
    if let Some(start_quote) = line.find('"')
    {
        if let Some(end_quote) = rust::find_start_at(line, start_quote+1, '"')
        {
            line[end_quote+1..].to_string()
        }
        else
        {
            "".to_string()
        }
    }
    else
    {
        "".to_string()
    }
}


fn get_text_after_include(line: &str) -> String
{
    if let Some(angle) = line.find('>')
    {
        if let Some(start_quote) = line.find('"')
        {
            if start_quote < angle
            {
                return get_text_after_include_quote(line);
            }
        }

        line[angle+1..].to_string()
    }
    else
    {
        get_text_after_include_quote(line)
    }
}


struct ClassifiedFile
{
    first_line: Option<usize>,
    last_line: Option<usize>,
    includes: Vec<Include>,
    invalid_order: bool,
}


fn classify_file
(
    lines: &[String],
    missing_files: &mut HashSet<String>,
    print: &mut printer::Printer,
    data: &builddata::BuildData,
    filename: &Path,
    verbose: bool,
    print_include_order_error_for_include: bool,
    include_error_message: MessageType
) -> Option<ClassifiedFile>
{

    let mut r = ClassifiedFile
    {
        first_line: None,
        last_line: None,
        includes: Vec::<Include>::new(),
        invalid_order: false
    };

    let mut last_class = -1;
    let mut line_num = 0;
    for line in lines
    {
        line_num += 1;
        
        if line.starts_with("#include")
        {
            if r.first_line.is_some() == false
            {
                r.first_line = Some(line_num);
            }
            r.last_line = Some(line_num);
            let l = line.trim_end();
            let line_class = classify_line(missing_files, print, data, l, filename, line_num)?;
            r.includes.push(Include::new(line_class, l));
            if last_class > line_class
            {
                if print_include_order_error_for_include
                {
                    print_message(include_error_message, print, filename, line_num, format!("Include order error for {}", l).as_str());
                }
                r.invalid_order = true;
            }
            last_class = line_class;
            if verbose
            {
                print.info(format!("{} {}", line_class, l).as_str());
            }
        }
    }

    r.includes.sort_by(include_compare);

    Some(r)
}


fn can_fix_and_print_errors
(
    lines: &[String],
    f: &ClassifiedFile,
    print: &mut printer::Printer,
    filename: &Path,
    first_error_only: bool
) -> bool
{
    let mut ok = true;

    if let (Some(first_line_found), Some(last_line_found)) = (f.first_line, f.last_line)
    {
        for line_num in first_line_found .. last_line_found
        {
            let print_this_error = match (ok, first_error_only)
            {
                // this is not the first error AND we only want to print the first error
                (false, true) => false,
                _ => true
            };

            let line = lines[line_num-1].trim();
            if line.is_empty()
            {
                // ignore empty
            }
            else if line.starts_with("#include")
            {
                let end_not_trimmed = get_text_after_include(line);
                let end = end_not_trimmed.trim();
                if end.is_empty() == false && end.starts_with("//") == false
                {
                    if print_this_error
                    {
                        error(print, filename, line_num, format!("Invalid text after include: {}", end).as_str());
                    }
                    ok = false;
                }
            }
            else
            {
                if print_this_error
                {
                    error(print, filename, line_num, format!("Invalid line {}", line).as_str());
                }
                ok = false;
            }
        }
    }

    ok
}


fn generate_suggested_include_lines_from_sorted_includes(includes: &[Include]) -> Vec<String>
{
    let mut new_lines = Vec::<String>::new();
    let mut current_class = includes[0].line_class;
    for i in includes
    {
        if current_class != i.line_class
        {
            new_lines.push("".to_string());
        }
        current_class = i.line_class;
        
        new_lines.push(i.line.to_string());
    }

    new_lines
}


fn compose_new_file_content(first_line_found: usize, last_line_found: usize, new_lines: &[String], lines: &[String]) -> Vec::<String>
{
    let mut file_data = Vec::<String>::new();
    
    for line_num in 1 .. first_line_found
    {
        file_data.push(lines[line_num-1].to_string());
    }
    for line in new_lines
    {
        file_data.push(line.to_string());
    }
    for line_num in last_line_found+1 .. lines.len()+1
    {
        file_data.push(lines[line_num-1].to_string());
    }

    file_data
}


fn print_lines(print: &mut printer::Printer, lines: &[String])
{
    print.info("*************************************************");
    for line in lines
    {
        print.info(line);
    }
    print.info("*************************************************");
    print.info("");
    print.info("");
}


enum Command
{
    MissingPatterns,
    ListUnfixable
    {
        print_first_error_only: bool
    },
    Check,
    Fix
    {
        nop: bool
    }
}


fn run_file
(
    missing_files: &mut HashSet<String>,
    print: &mut printer::Printer,
    data: &builddata::BuildData,
    verbose: bool,
    filename: &Path,
    command: &Command,
) -> bool
{
    let command_is_list_unfixable = match command { Command::ListUnfixable{print_first_error_only:_} => true, _ => false };
    let command_is_check          = match command { Command::Check                                   => true, _ => false };
    let command_is_fix            = match command { Command::Fix{nop:_}                              => true, _ => false };

    let print_include_order_error_for_include = command_is_check || command_is_fix;

    if verbose
    {
        print.info(format!("Opening file {}", filename.display()).as_str());
    }
    
    let loaded_lines = core::read_file_to_lines(filename);
    if let Err(err) = loaded_lines
    {
        print.error(format!("Failed to load {}: {}", filename.display(), err).as_str());
        return false;
    }

    let lines = loaded_lines.unwrap();
    let classified = match classify_file
    (
        &lines,
        missing_files,
        print,
        data,
        filename,
        verbose,
        print_include_order_error_for_include,
        if command_is_fix { MessageType::Warning } else { MessageType::Error }
    )
    {
        Some(c) => c,
        None =>
        {
            // file contains unclassified header
            return false;
        }
    };

    if classified.invalid_order == false
    {
        // this file is ok, don't touch it
        return true;
    }

    // if the include order is invalid, that means there needs to be a include and we know the start and end of it
    let first_line = classified.first_line.unwrap();
    let last_line = classified.last_line.unwrap();

    if !(command_is_fix || command_is_check || command_is_list_unfixable)
    {
        // if we wan't to print the unmatched header we don't care about sorting the headers
        return true;
    }

    let print_first_error_only = match command
    {
        Command::Fix{nop:_} => true,
        Command::ListUnfixable{print_first_error_only} => *print_first_error_only,
        _ => false // don't care, shouldn't be possible
    };

    if can_fix_and_print_errors(&lines, &classified, print, filename, print_first_error_only) == false
    {
        if command_is_fix
        {
            // can't fix this file... error out
            return false;
        }
    }

    if command_is_list_unfixable
    {
        return true;
    }

    let sorted_include_lines = generate_suggested_include_lines_from_sorted_includes(&classified.includes);

    if let Command::Fix{nop} = command
    {
        let file_data = compose_new_file_content(first_line, last_line, &sorted_include_lines, &lines);

        if *nop
        {
            print.info(format!("Will write the following to {}", filename.display()).as_str());
            print_lines(print, &file_data);
        }
        else
        {
            core::write_string_to_file(print, filename, &file_data.join("\n"));
        }
    }
    else
    {
        print.info("I think the correct order would be:");
        print_lines(print, &sorted_include_lines);
    }

    true
}


fn common_main
(
    args: &MainData,
    print: &mut printer::Printer,
    data: &builddata::BuildData,
    command: &Command
)
{
    let mut error_count = 0;
    let mut file_count = 0;
    let mut file_error = 0;

    let mut missing_files = HashSet::<String>::new();

    for filename in &args.files
    {
        file_count += 1;
        let stored_error = error_count;

        let ok = run_file
        (
            &mut missing_files,
            print,
            data,
            args.verbose,
            filename,
            &command
        );

        if ok == false
        {
            error_count += 1;
        }

        if error_count != stored_error
        {
            file_error += 1;
        }
    }

    if args.status
    {
        print.info(format!("Files parsed: {}",  file_count).as_str());
        print.info(format!("Files errored: {}", file_error).as_str());
        print.info(format!("Errors found: {}",  error_count).as_str());
    }
}


pub fn main(print: &mut printer::Printer, data: &builddata::BuildData, args: &Options)
{
    match args
    {
        Options::ListUnfixable{main, all} => common_main(main, print, data, &Command::ListUnfixable{print_first_error_only: *all == false}),
        Options::Check{main}              => common_main(main, print, data, &Command::Check),
        Options::MissingPatterns{main}    => common_main(main, print, data, &Command::MissingPatterns),
        Options::Fix{main, write}         => common_main(main, print, data, &Command::Fix{nop: *write==false}),
    }
}

File: src/CheckIncludes.cs
using Spectre.Console;
using Spectre.Console.Cli;
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;

namespace Workbench.CheckIncludes;

internal class MainCommandSettings : CommandSettings
{
    [Description("Files to look at")]
    [CommandArgument(0, "<file>")]
    public string[] Files { get; set; } = Array.Empty<string>();

    [Description("Print general file status at the end")]
    [CommandOption("--status")]
    [DefaultValue(false)]
    public bool PrintStatusAtTheEnd { get; set; }

    [Description("Use verbose output")]
    [CommandOption("--verbose")]
    [DefaultValue(false)]
    public bool UseVerboseOutput { get; set; }
}


internal class Main
{
    internal static void Configure(IConfigurator config, string name)
    {
        config.AddBranch(name, cmake =>
        {
            cmake.SetDescription("Check the order of the #include statements");
            cmake.AddCommand<MissingPatternsCommand>("missing-patterns").WithDescription("Print headers that don't match any pattern so you can add more regexes");
            cmake.AddCommand<ListUnfixableCommand>("list-unfixable").WithDescription("Print headers that can't be fixed");
            cmake.AddCommand<CheckCommand>("check").WithDescription("Check for style errors and error out");
            cmake.AddCommand<FixCommand>("fix").WithDescription("Fix style errors and print unfixable");
        });
    }
}


internal sealed class MissingPatternsCommand : Command<MissingPatternsCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithLoadedBuildData
            (
                (print, data) => IncludeTools.CommonMain(settings, print, data, new Command.MissingPatterns())
            );
    }
}


internal sealed class ListUnfixableCommand : Command<ListUnfixableCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
        [Description("Print all errors per file, not just the first one")]
        [CommandOption("--all")]
        [DefaultValue(false)]
        public bool PrintAllErrors { get; set; }
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithLoadedBuildData
            (
                (print, data) => IncludeTools.CommonMain(settings, print, data, new Command.ListUnfixable(settings.PrintAllErrors == false))
            );
    }
}


internal sealed class CheckCommand : Command<CheckCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithLoadedBuildData
            (
                (print, data) => IncludeTools.CommonMain(settings, print, data, new Command.Check())
            );
    }
}


internal sealed class FixCommand : Command<FixCommand.Arg>
{
    public sealed class Arg : MainCommandSettings
    {
        [Description("Write fixes to file")]
        [CommandOption("--write")]
        [DefaultValue(false)]
        public bool WriteToFile { get; set; }
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Arg settings)
    {
        return CommonExecute.WithLoadedBuildData
            (
                (print, data) => IncludeTools.CommonMain(settings, print, data, new Command.Fix(settings.WriteToFile == false))
            );
    }
}


public class Include : IComparable<Include>
{
    public int LineClass {get;}
    public string Line {get;}

    public Include(int line_class, string line)
    {
        this.LineClass = line_class;
        this.Line = line;
    }

    public int CompareTo(Include? rhs)
    {
        if (rhs == null) return 1;

        var lhs = this;
        if (lhs.LineClass == rhs.LineClass)
        {
            return lhs.Line.CompareTo(rhs.Line);
        }
        else
        {
            return lhs.LineClass.CompareTo(rhs.LineClass);
        }
    }
}

public enum MessageType
{
    Error, Warning
}

public static class IncludeTools
{
    private static void PrintError(Printer print, string filename, int line, string message)
    {
        print.error($"{filename}({line}): error CHK3030: {message}");
    }

    private static void PrintWarning(Printer print, string filename, int line, string message)
    {
        print.warning($"{filename}({line}): warning CHK3030: {message}");
    }

    private static void PrintMessage(MessageType messageType, Printer print, string filename, int line, string message)
    {
        switch (messageType)
        {
            case MessageType.Error: PrintError(print, filename, line, message); break;
            case MessageType.Warning: PrintWarning(print, filename, line, message); break;
        }
    }

    public static TextReplacer CreateReplacer(string file_stem)
    {
        var replacer = new TextReplacer();
        replacer.Add("{file_stem}", file_stem);
        return replacer;
    }


    private static int? ClassifySingleLine
    (
        HashSet<string> missingFiles,
        Printer print,
        BuildData data,
        string line,
        string filename,
        int lineNumber
    )
    {
        var replacer = CreateReplacer(Path.GetFileNameWithoutExtension(filename));

        foreach (var (index, included_regex_group) in data.IncludeDirectories.Select(((value, i) => (i, value))))
        {
            foreach (var included_regex in included_regex_group)
            {
                var re = included_regex.GetRegex(print, replacer);

                if (re == null) { continue; }

                if (re.Matches(line) != null)
                {
                    return index;
                }
            }
        }

        if (missingFiles.Contains(line) == false)
        {
            missingFiles.Add(line);
            PrintError(print, filename, lineNumber, $"{line} is a invalid header");
        }

        return null;
    }


    private static string GetTextAfterInclude(string line)
    {
        static string GetTextAfterQuote(string line)
        {
            var start_quote = line.IndexOf('"');
            if (start_quote == -1) { return ""; }

            var end_quote = line.IndexOf(line, start_quote + 1, '"');
            if (end_quote == -1) { return ""; }

            return line[(end_quote + 1)..];
        }

        var angle = line.IndexOf('>');
        if (angle == -1)
        {
            return GetTextAfterQuote(line);
        }

        var start_quote = line.IndexOf('"');

        if (start_quote != -1)
        {
            if (start_quote < angle)
            {
                return GetTextAfterQuote(line);
            }
        }

        return line[(angle + 1)..];
    }


    private static ClassifiedFile ClassifyFile
    (
        IEnumerable<string> lines,
        HashSet<string> missing_files,
        Printer print,
        BuildData data,
        string filename,
        bool verbose,
        bool print_include_order_error_for_include,
        MessageType include_error_message
    )
    {
        var r = new ClassifiedFile();

        var last_class = -1;
        var line_num = 0;
        foreach (var line in lines)
        {
            line_num += 1;

            if (line.StartsWith("#include") == false) { continue; }

            if (r.firstLineIndex == null)
            {
                r.firstLineIndex = line_num;
            }
            r.lastLineIndex = line_num;

            var l = line.TrimEnd();
            var line_class = ClassifySingleLine(missing_files, print, data, l, filename, line_num);
            if (line_class == null) { continue; }
            r.Includes.Add(new Include(line_class.Value, l));
            if (last_class > line_class.Value)
            {
                if (print_include_order_error_for_include)
                {
                    PrintMessage(include_error_message, print, filename, line_num, $"Include order error for {l}");
                }
                r.HasInvalidOrder = true;
            }
            last_class = line_class.Value;
            if (verbose)
            {
                print.Info($"{line_class.Value} {l}");
            }
        }

        r.Includes.Sort();
        return r;
    }

    private static bool CanFixAndPrintErrors
    (
        string[] lines,
        ClassifiedFile f,
        Printer print,
        string filename,
        bool printFirstErrorOnly
    )
    {
        if (f.firstLineIndex == null || f.lastLineIndex == null)
        {
            return true;
        }

        var ok = true;

        var first_line_found = f.firstLineIndex.Value;
        var last_line_found = f.lastLineIndex.Value;

        for (var line_num = first_line_found; line_num < last_line_found; line_num += 1)
        {
            var print_this_error = (ok, printFirstErrorOnly) switch
            {
                // this is not the first error AND we only want to print the first error
                (false, true) => false,

                // otherwise, print the error
                _ => true
            };

            var line = lines[line_num - 1].Trim();

            // ignore empty
            if (string.IsNullOrEmpty(line)) { continue; }

            if (line.StartsWith("#include"))
            {
                var end = GetTextAfterInclude(line).Trim();
                if (string.IsNullOrEmpty(end) == false && end.StartsWith("//") == false)
                {
                    if (print_this_error)
                    {
                        PrintError(print, filename, line_num, $"Invalid text after include: {end}");
                    }
                    ok = false;
                }
            }
            else
            {
                if (print_this_error)
                {
                    PrintError(print, filename, line_num, $"Invalid line {line}");
                }
                ok = false;
            }
        }

        return ok;
    }

    private static IEnumerable<string> GenerateSuggestedIncludeLinesFromSortedIncludes(IEnumerable<Include> includes)
    {
        int? current_class = null;
        foreach (var i in includes)
        {
            if (current_class == null)
            {
                current_class = i.LineClass;
            }

            if (current_class.Value != i.LineClass)
            {
                yield return "";
            }
            current_class = i.LineClass;

            yield return i.Line;
        }
    }


    private static IEnumerable<string> compose_new_file_content(int first_line_found, int last_line_found, string[] new_lines, string[] lines)
    {
        for (int line_num = 1; line_num < first_line_found; line_num++)
        {
            yield return lines[line_num - 1];
        }
        foreach (var line in new_lines)
        {
            yield return line;
        }
        for (int line_num = last_line_found + 1; line_num < lines.Length + 1; line_num += 1)
        {
            yield return lines[line_num - 1];
        }
    }

    private static void print_lines(Printer print, IEnumerable<string> lines)
    {
        print.Info("*************************************************");
        foreach (var line in lines)
        {
            print.Info(line);
        }
        print.Info("*************************************************");
        print.Info("");
        print.Info("");
    }


    private static bool RunFile
    (
        HashSet<string> missingFiles,
        Printer print,
        BuildData data,
        bool verbose,
        string filename,
        Command command
    )
    {
        var command_is_list_unfixable = command is Command.ListUnfixable;
        var command_is_check = command is Command.Check;
        var command_is_fix = command is Command.Fix;

        var print_include_order_error_for_include = command_is_check || command_is_fix;

        if (verbose)
        {
            print.Info($"Opening file {filename}");
        }

        var lines = Core.ReadFileToLines(filename);
        if (lines == null)
        {
            print.error($"Failed to load {filename}");
            return false;
        }

        var classified = ClassifyFile
        (
            lines,
            missingFiles,
            print,
            data,
            filename,
            verbose,
            print_include_order_error_for_include,
            command_is_fix ? MessageType.Warning : MessageType.Error
        );
        if (classified == null)
        {
            // file contains unclassified header
            return false;
        };

        if (classified.HasInvalidOrder == false)
        {
            // this file is ok, don't touch it
            return true;
        }

        if (classified.firstLineIndex == null || classified.lastLineIndex == null) { throw new Exception("bug!"); }
        // if the include order is invalid, that means there needs to be a include and we know the start and end of it
        var firstLine = classified.firstLineIndex.Value;
        var lastLine = classified.lastLineIndex.Value;

        if (!(command_is_fix || command_is_check || command_is_list_unfixable))
        {
            // if we wan't to print the unmatched header we don't care about sorting the headers
            return true;
        }

        bool printFirstErrorOnly = command switch
        {
            Command.Fix => true,
            Command.ListUnfixable p => p.printFirstErrorOnly,
            _ => false // don't care, shouldn't be possible
        };

        if (CanFixAndPrintErrors(lines, classified, print, filename, printFirstErrorOnly) == false && command_is_fix)
        {
            // can't fix this file... error out
            return false;
        }

        if (command_is_list_unfixable)
        {
            return true;
        }

        var sortedIncludeLines = GenerateSuggestedIncludeLinesFromSortedIncludes(classified.Includes).ToArray();

        switch (command)
        {
            case Command.Fix nop:
                var fileData = compose_new_file_content(firstLine, lastLine, sortedIncludeLines, lines);

                if (nop.nop)
                {
                    print.Info($"Will write the following to {filename}");
                    print_lines(print, fileData);
                }
                else
                {
                    File.WriteAllLines(filename, fileData);
                }
                break;
            default:
                print.Info("I think the correct order would be:");
                print_lines(print, sortedIncludeLines);
                break;
        }

        return true;
    }

    internal static int CommonMain
    (
        MainCommandSettings args,
        Printer print,
        BuildData data,
        Command command
    )
    {
        var errorCount = 0;
        var fileCount = 0;
        var fileError = 0;

        var missingFiles = new HashSet<string>();

        foreach (var filename in args.Files)
        {
            fileCount += 1;
            var storedError = errorCount;

            var ok = RunFile
            (
                missingFiles,
                print,
                data,
                args.UseVerboseOutput,
                filename,
                command
            );

            if (ok == false)
            {
                errorCount += 1;
            }

            if (errorCount != storedError)
            {
                fileError += 1;
            }
        }

        if (args.PrintStatusAtTheEnd)
        {
            print.Info($"Files parsed: {fileCount}");
            print.Info($"Files errored: {fileError}");
            print.Info($"Errors found: {errorCount}");
        }

        return errorCount;
    }

}

public class ClassifiedFile
{
    public int? firstLineIndex{get; set;}
    public int? lastLineIndex{get; set;}
    public List<Include> Includes { get; } = new();
    public bool HasInvalidOrder { get; set; } = false;
}


public abstract record Command
{
    public record MissingPatterns() : Command;
    public record ListUnfixable(bool printFirstErrorOnly) : Command;
    public record Check() : Command;
    public record Fix(bool nop) : Command;
}

